server/lib/scheduler/Scheduler.ts

```diff
@@ -0,0 +1,110 @@
+    return assignment.getWorkspaceId();
+  }
+
+  public async findNextWorkspace(userId) {
+    // clear cache so we don't use old eligibility info
+    this.rootParentCache.clearRootParentCache();
+    const idsOfWorkspacesThatCouldBeNext = await this.getIdsOfWorkspacesThatCouldBeNext(userId);
+    const assignedWorkspaceId = this.pickWorkspaceIdAtRandom(idsOfWorkspacesThatCouldBeNext);
+    await this.schedule.assignWorkspaceToUser(userId, assignedWorkspaceId);
+  }
+
+  private async getIdsOfWorkspacesThatCouldBeNext(userId) {
+    const allWorkspaces = await this.fetchAllWorkspaces();
+    const allEligibleWorkspaces = await this.filterByEligibility(allWorkspaces);
+    const workspacesInTreeWorkedOnLeastRecently = await this.filterByWhetherInTreeWorkedOnLeastRecently(allEligibleWorkspaces);
+    const notYetWorkedOnInThatTree = await this.filterByWhetherNotYetWorkedOn(workspacesInTreeWorkedOnLeastRecently);
+
+    let finalWorkspaces = notYetWorkedOnInThatTree;
+
+    // if every workspace in that tree has been worked on
+    // then instead look for workspaces with remaining budgets
+    if (finalWorkspaces.length === 0) {
+      finalWorkspaces = await this.filterByWhetherHasRemainingBudget(workspacesInTreeWorkedOnLeastRecently);
+    }
+
+    // if every workspace in that tree has 0 remaining budget
+    // then instead look for the workspace worked on least recently
+    if (finalWorkspaces.length === 0) {
+      const idOfWorkspaceWorkedOnLeastRecently = await this.schedule.getLeastRecentlyActiveWorkspace(workspacesInTreeWorkedOnLeastRecently.map(w => w.id));
+      return [idOfWorkspaceWorkedOnLeastRecently];
+    }
+
+    return finalWorkspaces.map(w => w.id);
+  }
+
+  private pickWorkspaceIdAtRandom(workspaceIds) {
+    const randomIndex = Math.floor(Math.random() * (workspaceIds.length));
+    return workspaceIds[randomIndex];
+  }
+
+  private async filterByEligibility(workspaces) {
+    // use for... of instead of asyncro's filter
+    // because isWorkspaceEligible will use rootParentCache
+    // and the parallel nature of filter doesn't work with
+    // rootParentCache's caching
+    let allEligibleWorkspaces = [];
+    for (const workspace of workspaces) {
+      const isMarkedEligible = await this.isWorkspaceEligible(workspace);
+      const isCurrentlyBeingWorkedOn = await this.schedule.isWorkspaceCurrentlyBeingWorkedOn(workspace.id);
+      if (isMarkedEligible && !isCurrentlyBeingWorkedOn) {
+        allEligibleWorkspaces.push(workspace);
+      }
+    }
+    return allEligibleWorkspaces;
+  }
+
+  /*
+    NOTE: this only considers the trees that at least one workspace in the
+    workspaces argument belong to
+  */
+  private async filterByWhetherInTreeWorkedOnLeastRecently(workspaces) {
+    let workspacesInTreeWorkedOnLeastRecently = [];
+
+    for (const workspace of workspaces) {
+      const shouldInclude = await this.schedule.isInTreeWorkedOnLeastRecently(workspaces.map(w => w.id), workspace.id);
```

Comment:
If this is needed, it would make sense to me for ``this.schedule`` to expose the full list of workedOnRecently workspaces, and then this could be a simple comparison filter.